### ISSUE TITLE
update test cases to LLVM 9.0

### ISSUE THREAD

**Issue by mewmew** [llir/llvm#105]
Some of the previously valid test cases (taken from the 7.0 release of LLVM) are no longer valid.

Error reported by `opt` of the LLVM 9.0 release when parsing [testdata/llvm/test/Analysis/DominanceFrontier/new_pm_test.ll](https://github.com/llir/testdata/blob/7bf108448acdfe95ef22f20b66335b767e10626a/llvm/test/Analysis/DominanceFrontier/new_pm_test.ll):
```bash
$ opt -S -o foo.ll new_pm_test.ll
opt: new_pm_test.ll:29:1: error: label expected to be numbered '12'
13:
^
```

Error reported by `go test` for `llir/llvm` once updated to handle explicitly named unnamed basic blocks (as done in LLVM 9.0); see #104.

From https://travis-ci.org/llir/llvm/jobs/605020808#L708:

```
--- FAIL: TestParseFile (40.39s)

    asm_test.go:533: unable to parse "../testdata/llvm/test/Analysis/DominanceFrontier/new_pm_test.ll" into AST;
    invalid local ID in function "@a_linear_impl_fig_1", expected %12, got %13
```

**Comment by mewmew** [comment]
The `test/Analysis/DominanceFrontier/new_pm_test.ll` test case was updated from LLVM 7.0 to LLVM 9.0, as follows:

```diff
--- a/llvm/test/Analysis/DominanceFrontier/new_pm_test.ll
+++ b/llvm/test/Analysis/DominanceFrontier/new_pm_test.ll
@@ -3,48 +3,48 @@
 
 define void @a_linear_impl_fig_1() nounwind {
 0:
-  br label %"1"
+  br label %1
 1:
-  br label %"2"
+  br label %2
 2:
-  br label %"3"
+  br label %3
 3:
-  br i1 1, label %"13", label %"4"
+  br i1 1, label %12, label %4
 4:
-  br i1 1, label %"5", label %"1"
+  br i1 1, label %5, label %1
 5:
-  br i1 1, label %"8", label %"6"
+  br i1 1, label %8, label %6
 6:
-  br i1 1, label %"7", label %"4"
+  br i1 1, label %7, label %4
 7:
   ret void
 8:
-  br i1 1, label %"9", label %"1"
+  br i1 1, label %9, label %1
 9:
-  br label %"10"
+  br label %10
 10:
-  br i1 1, label %"12", label %"11"
+  br i1 1, label %13, label %11
 11:
-  br i1 1, label %"9", label %"8"
-13:
-  br i1 1, label %"2", label %"1"
+  br i1 1, label %9, label %8
 12:
-   switch i32 0, label %"1" [ i32 0, label %"9"
-                              i32 1, label %"8"]
+  br i1 1, label %2, label %1
+13:
+   switch i32 0, label %1 [ i32 0, label %9
+                              i32 1, label %8]
 }
```

**Comment by mewmew** [comment]
Note, to update the test cases from LLVM 7.0 to LLVM 9.0, the following steps were taken.

Update VER in [testdata/llvm/Makefile](https://github.com/llir/testdata/blob/master/llvm/Makefile):

```diff
 # LLVM version.
-VER=7.0.0
+VER=9.0.0
```

Run `make` to download the LLVM 9.0 source release and extract `*.ll` test cases from it.

From the `testdata/llvm` directory of the [testdata](https://github.com/llir/testdata) repository.
```bash
make
```

Run [skip.sh](https://github.com/llir/testdata/blob/master/llvm/skip.sh) to skip a few "bad" test cases.

```bash
./skip.sh
```

Commit new test cases to the Git repo.

```
git add llvm/test
git commit -m 'llvm: update test cases to LLVM 9.0 release'
```

The remaining step is to update the `testdata` submodule of the `llir/llvm` repository, and then run `go test github.com/llir/llvm/...` to see if any of the *golden* output changed.

Each test case (e.g. [llvm/test/Analysis/DominanceFrontier/new_pm_test.ll](https://github.com/llir/testdata/blob/master/llvm/test/Analysis/DominanceFrontier/new_pm_test.ll)) has a golden output (e.g. [llvm/test/Analysis/DominanceFrontier/new_pm_test.ll.golden](https://github.com/llir/testdata/blob/master/llvm/test/Analysis/DominanceFrontier/new_pm_test.ll.golden)).

Sometimes the expected output changes when the grammar of LLVM IR is updated in between LLVM releases. Otherwise the expected output should remain stable.

Once the test cases are commited, we enter an iterative process of updating the [LLVM IR grammar](https://github.com/llir/ll) to incorporate new changes made to the LLVM IR language as part of the release, and then if new concepts are added to the language, then add the respective concepts to `llir/llvm/ir`. For the LLVM 9.0 release, this corresponds to the tracking issue #101.

**Comment by mewmew** [comment]
The only test case currently failing is `testdata/llvm/test/Assembler/diexpression.ll`, because the parser generated from the [ll.tm](!7 = !DIExpression(DW_OP_LLVM_convert, 16, DW_ATE_unsigned, DW_OP_LLVM_convert, 32, DW_ATE_signed)) grammar fails to parse this line:

```llvm
!7 = !DIExpression(DW_OP_LLVM_convert, 16, DW_ATE_unsigned, DW_OP_LLVM_convert, 32, DW_ATE_signed)
```

Results from `go test ./...`
```
--- FAIL: TestParseFile (34.42s)
    asm_test.go:533: unable to parse "../testdata/llvm/test/Assembler/diexpression.ll" into AST; syntax error at line 24
        unable to parse "../testdata/llvm/test/Assembler/diexpression.ll" into an AST
        github.com/llir/llvm/asm.ParseString
        	/home/u/Desktop/go/src/github.com/llir/llvm/asm/asm.go:58
        github.com/llir/llvm/asm.ParseBytes
        	/home/u/Desktop/go/src/github.com/llir/llvm/asm/asm.go:48
        github.com/llir/llvm/asm.ParseFile
        	/home/u/Desktop/go/src/github.com/llir/llvm/asm/asm.go:29
        github.com/llir/llvm/asm.TestParseFile
        	/home/u/Desktop/go/src/github.com/llir/llvm/asm/asm_test.go:531
        testing.tRunner
        	/home/u/go/src/testing/testing.go:954
        runtime.goexit
        	/home/u/go/src/runtime/asm_amd64.s:1375
```

**Comment by mewmew** [comment]
@dannypsnl, feel like taking a look at adding the enums needed to help the parser handle this line? (I think it is `DW_OP_LLVM_xxx`)

```llvm
!7 = !DIExpression(DW_OP_LLVM_convert, 16, DW_ATE_unsigned, DW_OP_LLVM_convert, 32, DW_ATE_signed)
```

**Comment by dannypsnl** [comment]
I dig into this case, the parser returns error: `unable to locate DwarfOp enum corresponding to "DW_OP_LLVM_convert"`, but it seems like there has rule for it?

https://github.com/llir/grammar/blob/f1eb80e86b542bd8a463371e8a99e116c61d2364/ll.tm#L107

**Comment by dannypsnl** [comment]
By the way, would you like to use official IR Parser by submodule LLVM-IR-parser and mapping ast only? [textmapper](https://github.com/inspirer/textmapper) didn't provide a reasonable result for parsing error. I didn't know which step it stuck.

**Comment by mewmew** [comment]
Hi @dannypsnl!

Great that you started working on this!

The test case you wrote is good. Unfortunately Textmapper does not show column, but only line offset of errors. I'll submit an issue upstream.

One thing you could do for now to get more useful errors is to split the line into multiple lines (since whitespace tokens are ignored):

```llvm
!7
=
!DIExpression(DW_OP_LLVM_convert,
16,
DW_ATE_unsigned,
DW_OP_LLVM_convert,
32,
DW_ATE_signed
)
```

Now, we get the error at "line 5".

```bash
$ go test -v
=== RUN   TestParse
    TestParse: aaa_test.go:18: syntax error at line 5
```

Which corresponds to this line:

```llvm
DW_ATE_unsigned,
```

So, we now know it fails to parse Dwarf Attribute Encodings in `DIExpression`s. Lets look at the update to the C++ parser in LLVM 9.0 to see what changed.

An update was made to `ParseDIExpression` in LLVM 9.0. It now includes `DwarfAttEncoding` as a valid field. Previously only `DwarfOp` and `Int` were valid.

To see the diff, run:

```bash
diff -u llvm-7.0.0.src/lib/AsmParser/ llvm-9.0.0.src/lib/AsmParser/
```

The relevant part for this change is in `lib/AsmParser/LLParser.cpp`, more specifically in `LLParser::ParseDIExpression`, where the following code was added:

```diff
+      if (Lex.getKind() == lltok::DwarfAttEncoding) {
+        if (unsigned Op = dwarf::getAttributeEncoding(Lex.getStrVal())) {
+          Lex.Lex();
+          Elements.push_back(Op);
+          continue;
+        }
+        return TokError(Twine("invalid DWARF attribute encoding '") + Lex.getStrVal() + "'");
+      }
+
```

So, the `ll.tm` grammar has to be updated accordingly.

From:
```
DIExpressionField -> DIExpressionField
	: UintLit
	| DwarfOp
;
```

To:
```
DIExpressionField -> DIExpressionField
	: UintLit
	| DwarfAttEncoding
	| DwarfOp
;
```

**Comment by mewmew** [comment]
The grammar has now been updated https://github.com/llir/grammar/commit/7cadd4c7729f093d0187d77566c6479fd69c7dd4

**Comment by mewmew** [comment]
Fixed by @dannypsnl in #110.

**Comment by mewmew** [comment]
> The test case you wrote is good. Unfortunately Textmapper does not show column, but only line offset of errors. I'll submit an issue upstream.

Reported upstream at https://github.com/inspirer/textmapper/issues/36
